### PR TITLE
Backport "Implement Freeform Unmanaged Error" to 2018-04

### PR DESCRIPTION
**Pull request by alexanderkyte**

### CODE REVIEW

File: mono/eglib/goutput.c [hunk 1]
```diff
@@ -126,12 +126,23 @@ g_log (const gchar *log_domain, GLogLevelFlags log_level, const gchar *format, .
...
+}
+
 void
 g_assertion_message (const gchar *format, ...)
 {
 	va_list args;
 
 	va_start (args, format);
+
+	g_vasprintf (&failure_assertion, format, args);
```
Comment (jaykrell) [inline mono/eglib/goutput.c]: This breaks assert. It gets varargs after g_vasprintf has bumped it past the parameters.
You get garbage or a crash in the assertion failure.
Put va_start after it, or g_vasprintf into a local and use it for both (g_log("%s", local)).

Comment (jaykrell) [reply mono/eglib/goutput.c]: Adding va_start is small change, reusing sprintf output is "faster", and maybe feels safer -- is reusing varargs portable with va_start?

File: mono/eglib/glib.h [hunk 2]
```diff
@@ -584,6 +584,8 @@ GLogLevelFlags g_log_set_fatal_mask   (const gchar *log_domain, GLogLevelFlags f
 void           g_logv                 (const gchar *log_domain, GLogLevelFlags log_level, const gchar *format, va_list args);
 void           g_log                  (const gchar *log_domain, GLogLevelFlags log_level, const gchar *format, ...);
 void           g_assertion_message    (const gchar *format, ...) G_GNUC_NORETURN;
+void           g_assertion_message    (const gchar *format, ...) G_GNUC_NORETURN;
```
Comment (jaykrell) [inline mono/eglib/glib.h]: Why the extra declaration?

File: mono/eglib/goutput.c (hunk 1 above)
Comment (jaykrell) [reply mono/eglib/goutput.c]: https://github.com/mono/mono/pull/8727/commits/a1baf487d073f6d66d11ac0134f2fa6b0e06408e fixes this.

**Comment by alexanderkyte** [comment]
The test failure that blocks merging is: 
````03:34:57    0 test(s) timed out
03:34:57 
03:34:57 Failed test(s):
03:34:57 
03:34:57 sgen-new-threads-dont-join-stw.exe
03:34:57 =============== sgen-new-threads-dont-join-stw.exe.stdout ===============
03:34:57 ........
03:34:57 done 1 iterations
03:34:57 
* Assertion at unwind.c:640, condition `cfa_reg != -1' not met
03:34:57 
03:34:57 Waiting for signalled thread to collect stacktrace
03:34:57 Waiting for signalled thread to collect stacktrace
03:34:57 
03:34:57 =============== EOF ===============
03:34:57 =============== sgen-new-threads-dont-join-stw.exe.stderr ===============
03:34:57 Stacktrace:
03:34:57 
03:34:57 
03:34:57 
Native stacktrace:
03:34:57 
03:34:57 	0   mono                                0x0011a499 mono_handle_native_crash + 233
03:34:57 	1   mono                                0x00180ec9 sigabrt_signal_handler + 153
03:34:57 	2   libsystem_platform.dylib            0xa1738e5b _sigtramp + 43
03:34:57 	3   ???                                 0xffffffff 0x0 + 4294967295
03:34:57 	4   libsystem_c.dylib                   0xa15bea34 abort + 141
03:34:57 	5   mono                                0x0031b733 mono_log_write_logfile + 355
03:34:57 	6   mono                                0x003166a1 structured_log_adapter + 33
03:34:57 	7   mono                                0x00333810 monoeg_g_logv + 80
03:34:57 	8   mono                                0x00333979 monoeg_assertion_message + 57
03:34:57 	9   mono                                0x00133664 mono_unwind_frame + 1780
03:34:57 	10  mono                                0x0017de1f mono_arch_unwind_frame + 319
03:34:57 	11  mono                                0x00117392 arch_unwind_frame + 50
03:34:57 	12  mono                                0x001170db mono_find_jit_info_ext + 203
03:34:57 	13  mono                                0x00118233 unwinder_unwind_frame + 195
03:34:57 	14  mono                                0x00118dab mono_handle_exception_internal + 1947
03:34:57 	15  mono                                0x001185fa mono_handle_exception + 42
03:34:57 	16  mono                                0x001168b9 mono_raise_exception_with_ctx + 25
03:34:57 	17  mono                                0x00270075 self_interrupt_thread + 69
03:34:57 	18  ???                                 0x00518f09 0x0 + 5345033
03:34:57 	19  ???                                 0x0056d0dc 0x0 + 5689564
03:34:57 	20  ???                                 0x0056cdfc 0x0 + 5688828
03:34:57 	21  ???                                 0x0056cc48 0x0 + 5688392
03:34:57 	22  ???                                 0x0056cb1c 0x0 + 5688092
03:34:57 	23  ???                                 0x0056793c 0x0 + 5667132
03:34:57 	24  mscorlib.dll.dylib                  0x01a5e355 System_Threading_Timer_Scheduler_TimerCB_object + 309
03:34:57 	25  mscorlib.dll.dylib                  0x01a54e18 System_Threading_QueueUserWorkItemCallback_System_Threading_IThreadPoolWorkItem_ExecuteWorkItem + 56
03:34:57 	26  mscorlib.dll.dylib                  0x01a5344b System_Threading_ThreadPoolWorkQueue_Dispatch + 555
03:34:57 	27  mscorlib.dll.dylib                  0x01a54c58 System_Threading__ThreadPoolWaitCallback_PerformWaitCallback + 24
03:34:57 	28  ???                                 0x0056760d 0x0 + 5666317
03:34:57 	29  mono                                0x0006f9db mono_jit_runtime_invoke + 843
03:34:57 	30  mono                                0x0024af8a do_runtime_invoke + 74
03:34:57 	31  mono                                0x00248258 mono_runtime_try_invoke + 88
03:34:57 	32  mono                                0x00271df2 worker_callback + 930
03:34:57 	33  mono                                0x001be4a1 worker_thread + 417
03:34:57 	34  mono                                0x0026f67a start_wrapper + 682
03:34:57 	35  libsystem_pthread.dylib             0xa1742047 _pthread_body + 184
03:34:57 	36  libsystem_pthread.dylib             0xa1741f8f _pthread_body + 0
03:34:57 	37  libsystem_pthread.dylib             0xa174184a thread_start + 34```

**Pull request merged**